HIP/hip-904-frictionless-airdrop.md

```diff
@@ -0,0 +1,714 @@
+*/ 
+message TokenAirdropTransactionBody {
+    /**
+    * The desired list of tokens to airdrop. If any custom fees are assessed,
+    * the ledger will try to deduct them from the payer of this airdrop,
+    * resolving the transaction to INSUFFICIENT_PAYER_BALANCE_FOR_CUSTOM_FEE if
+    * this is not possible.
+    * The sender must have all the succifient balance at the time of the transaction to support this transaction.
+    *
+    * If the receiver account has already associated with the given tokens, then
+    * the fee to transfer will be identical to a token transfer. If the receiver
+    * has open auto-association slots but has not yet associated with the token,
+    * then the payer will pay for the transfer, association, one renewal credit,
+    * and airdrop fee. If the receiver has not associated with the token and has
+    * no open association slot, then the transfer will not deliver the token,
+    * but will create a `pending airdrop` instead, with the receiver account
+    * listed on the `pending airdrop`. The payer will pay for the transfer,
+    * association, one renewal credit, and airdrop fee.
+    * All tokenTransferList items must succeed (transfer of pedning state) for the transaction to succeed.
+    */
+    repeated TokenTransferList tokenTransfers = 1;
+}
+```
+
+TokenAirdrop transaction will support the transfers of value in a similar manner to CryptoTransfer with the difference of only supporting tokens (no HBAR transfer support) and supporting the resulting pending transfer state where a CryptoTransfer would have failed due to the lack of association options.
+
+However, it is important to have a separate HAPI transaction type for a few reasons
+
+- the ability to call out the potential separate behaviours based on maxAutomaticAssociation, whereby transfer can end up in a pending transfer state or a consumption of an association slot.
+- cleaner conceptual understanding
+- naming allows the grouping of functionality with Cancel and Claim which complete the pending state a TokenAirdrop may creates
+- explicit specification in record stream.
+- improved developer experience
+
+Similar restrictions around transfer list and aggregation across sender and receiver from CryptoTransfer will be applied to TokenAirdop e.g. max 10 transfers per transaction
+
+At the time of this HIP `TokenAirdropTransaction` will not be supported by the Scheduled Transaction Service. A future HIP may explore this support.
+
+#### Cancel
+
+```protobuf
+message TokenPendingAirdrop {
+    /**
+    * Account that issue the airdrop trasnaction
+    */
+	AccountID sender_id = 1;
+
+	/**
+    * Account that was the intended reciever of the airdrop
+    */
+    AccountID receiver_id = 2;
+
+	/**
+    * The unique token identifier. If it's a FT then it's just the TokenID.
+	* If it's an NFT then NftID which captures both the TokenID and serial id
+    */
+	oneof token_id {
```

Comment:
So, if you do an airdrop of fungible token T from account S to account R, and it goes to pending, then you do _another_ one (in the same TokenAirdrop or a different one) and it too goes to pending, then when you cancel you cancel both of them, or IOW _all_ pending airdrops from S to R of T.  (Claim, which uses this same structure also.). Just checking.

Reply:
If T is fungible then yeah all of the pending amount will be cancelled.
If T is non-fungible then it's limited to 10 at a time to be cancelled or claimed